### ISSUE TITLE
Android Studio build

### ISSUE THREAD

**Issue by sasmaster** [kcat/openal-soft#132]
Well, this is a mess.I mean,anyone tried to build this lib with AndroidStudio?
I am trying using Clang toolchain.Getting this error:

.....AppData\Local\Temp\helpers-c437e5.s:573: Error: selected processor does not support ARM mode `vmsr fpscr,r0'

**Comment by kcat** [comment]
I played around a little bit with Android Studio, and I remember being able to build it by creating a new project (could probably use a preexisting one), and include OpenAL Soft in the project's main CMakeLists.txt using the ```add_subdirectory(...)``` command.

Building OpenAL Soft stand-alone would require a stand-alone toolchain, modifying and using the provided XCompile-Android.txt to build it. Though given how Android apps are built, I think including OpenAL Soft as described above would be more straight forward.

**Comment by sasmaster** [comment]
I tried exactly this setup.Id doesn't work.And inspecting the Cmake configs I see it was tested long ago on Android NDK toolchains. Since then even the naming of dirs in NDK had changed.And in my case I need to setup clang toolchain which is more tricky than GCC because it has llvm + GCC.... So far a lot of mess....

**Comment by sriharshachilakapati** [comment]
I got an older version (1.17.2 along with JNI bindings) available as a gradle dependency [here](https://github.com/sriharshachilakapati/AndroidOpenAL). I'm waiting to update to 1.18.0 because I heard there were some issues, especially there are some games getting broke with it.

Add this to the gradle dependencies:

~~~groovy
compile "com.goharsha:android-openal:0.1-SNAPSHOT"
~~~

You can find all the functions in the class `AL` as static functions.

**Comment by sriharshachilakapati** [comment]
@sasmaster I've updated the port to `0.3-SNAPSHOT` which contains version `1.18.0` the latest. It's working fine for me, I'll update the repo and make a proper release after testing all my other apps using it.

@kcat I'm thinking to change the versioning of my port to the same as the openalsoft versioning. Since I don't do any changes to the source code (except for committing pre-configured `config.h` and `version.h` files), having the same version as this original library makes it easier for people. What do you think?

**Comment by sasmaster** [comment]
Well,currently I am trying to build this repo manually.The VS complains regarding null.c .There are two null.c files in the codebase.Is it supposed to be like that?

**Comment by sasmaster** [comment]
I have managed to build this repo for ARM using Visual Studio 2015 Android native project.I feel like this is a hack but it works.Several moments I would like to point out for the repo maintainers:
1.Currently latest Android NDK,especially Clang toolchain just don't work with repo's Cmakes.I tried both Android Studio gradle/Cmake system and manual CMake config with Android's NDK and toolchains.

2.When trying to build the lib manually,no cmake,frankly speaking - it is a hell.First of all,version.h and config.h are Cmake generated headers.And while this seems to be quite nice practice for Cmake based projects,it set ton of problems if one tries to build this lib without CMake (yeah,that's possible,not everyone knows or wants to mess with CMake).

3.So I deleted version.h ref from everywhere in the source code.I also copy pasted config.h from Windows target generated Solution and started manually commenting and uncommenting related definitions.Which is,well,another hell,because no minimal comments provided on what flags should be set for what platform.So I hope I figured out the correct ones based on include files inspection and trial/error approach.

4.The repo contains code files with identical names.That's null.c which exists both in effect and Alc/backend dir. Now,here is a fanny thing - MSVC has no problem with that.But Clang does.It complains of multiple -o files which it cannot generate.I tried all sort of stuff,like configuring obj files dir so that each obj file goes into its relative dir etc... But nothing helped,so I finally solved that by renaming one of those null.c to 'null1.c'. And that worked.Now what is left to do is to build and test also ARM64.

**Comment by kcat** [comment]
> I tried exactly this setup.Id doesn't work.

The provided error doesn't unfortunately point to anything wrong in the code, since it's an error in its own generated assembly. What's the full output, from a fresh run of cmake to the error?

> @kcat I'm thinking to change the versioning of my port to the same as the openalsoft versioning. Since I don't do any changes to the source code (except for committing pre-configured config.h and version.h files), having the same version as this original library makes it easier for people. What do you think?

I don't see any problems with that.

> There are two null.c files in the codebase.Is it supposed to be like that?

Yes, they're in separate sub-directories (backend for the null backend, and effect for the null effect), so it's fine.

> 1.Currently latest Android NDK,especially Clang toolchain just don't work with repo's Cmakes.I tried both Android Studio gradle/Cmake system and manual CMake config with Android's NDK and toolchains.

What's the full output and error?

> 2.When trying to build the lib manually,no cmake,frankly speaking - it is a hell.

Unfortunately I haven't found any alternative build system that's as flexible and works on all potential targets (autotools is nearly impossible to use on Windows, and probably doesn't work on Android Studio either). It's far worse to have to maintain individual configurations for specific systems, since they'll easily get neglected and go out of date.

> 3.So I deleted version.h ref from everywhere in the source code.I also copy pasted config.h from Windows target generated Solution and started manually commenting and uncommenting related definitions.Which is,well,another hell,because no minimal comments provided on what flags should be set for what platform.

That's exactly what cmake is for, since not only is every platform is different, different versions of a given platforms can require (or work better with) different flags too. The detection done with cmake is far less of a hassle.

**Comment by scmcom** [comment]
Assuming you seek working .so binaries, this issue is relatively easily resolved by *not* using AS and referring to the following closed issue:
https://github.com/kcat/openal-soft/issues/46

Specifically this part:
"compiled with toolchain file: https://github.com/taka-no-me/android-cmake/blob/master/android.toolchain.cmake

cmake command line: cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/android.toolchain.cmake -DANDROID_NDK=~/Library/Android/sdk/ndk-bundle -DCMAKE_BUILD_TYPE=Release -DANDROID_ABI=armeabi -DANDROID_NATIVE_API_LEVEL=android-18 -DANDROID_STL=gnustl_static -DALSOFT_EMBED_HRTF_DATA=TRUE"

You need to use android cmake toolchain such as the one linked above with arguments such as these^, and working binaries will be generated. 

Simply put the android.toolchain.cmake in the OpenAL-Soft cmake dir and run the cmake command with the relevant args. No further changes to the repo's file system are needed.

**Comment by Slion** [comment]
I too am trying to build OpenAL from Android Studio.
I configured the CMake build as follow from gradle:
```
'-DALSOFT_NO_CONFIG_UTIL=ON',
'-DALSOFT_EXAMPLES=OFF',
'-DALSOFT_UTILS=OFF',
'-DALSOFT_TESTS=OFF'
```
Unfortunately the custom `native-tools` target failed complaining it could not find ninja, was it executing the proper cmake executable? I doubt it otherwise it should have found ninja.

I tried disabling the `native-tools` but then it fails later one saying it can't find `bsincgen` and others which are probably the missing native tools.

Questions:
Are those native tools needed to perform the Android build?
Is there no option to disable them?
I'm guessing they should be built for the host machine, Windows in my case rather than the target machine, Android in this case.
Shall I specify ALSOFT_NATIVE_TOOLS_PATH and use the ones that were built by my Windows build?

**Comment by kcat** [comment]
> Are those native tools needed to perform the Android build?

Yes, the native tools need to run to generate some headers that are included in the build (specifically, headers for the embedded HRTF file data, so HRTF can be used without external files, and a header for some large resampler filter tables, which are too big and complicated to keep updated by hand).

> I'm guessing they should be built for the host machine, Windows in my case rather than the target machine, Android in this case.
> Shall I specify ALSOFT_NATIVE_TOOLS_PATH and use the ones that were built by my Windows build?

Yes that should work, just make sure the Windows build is the same OpenAL Soft version as the one you're building for Android.

**Comment by Slion** [comment]
Looks like I'm one step further but now I'm getting the following errors. CLang and NDK 19.2.
Specifying the native tools path was not enough though.
I had to comment out the `ADD_CUSTOM_TARGET(native-tools` as it could not resolve its dependencies.
Now however I wonder how that works at all since other custom commands declare dependencies on `native-tools`.

```
 C:\Users\micro\AppData\Local\Android\sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=x86_64-none-linux-android21 --gcc-toolchain=C:/Users/micro/AppData/Local/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Users/micro/AppData/Local/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64/sysroot -fPIC -std=c++11 -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -Wa,--noexecstack -Wformat -Werror=format-security -stdlib=libc++ -fms-extensions -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -shared -Wl,-soname,libopenal.so -o C:\Dev\bin\Android\app\intermediates\cmake\debug\obj\x86_64\libopenal.so C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/common/alcomplex.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/common/alexcpt.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/common/almalloc.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/common/threads.cpp.o C:/Dev/bin//Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alAuxEffectSlot.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alBuffer.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alEffect.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alError.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alExtension.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alFilter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alListener.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alSource.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/alState.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/event.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/OpenAL32/sample_cvt.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/alc.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/alu.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/alconfig.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/bs2b.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/converter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mastering.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/ringbuffer.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/autowah.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/chorus.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/compressor.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/dedicated.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/distortion.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/echo.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/equalizer.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/fshifter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/modulator.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/null.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/pshifter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/reverb.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/filters/biquad.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/filters/nfc.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/filters/splitter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/helpers.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/hrtf.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/uhjfilter.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/ambdec.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/bformatdec.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/panning.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixvoice.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixer/mixer_c.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixer/mixer_sse.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixer/mixer_sse2.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixer/mixer_sse3.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/mixer/mixer_sse41.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/backends/base.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/backends/loopback.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/backends/null.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/backends/opensl.cpp.o C:/Dev/bin/Android/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/backends/wave.cpp.o  -pthread -lOpenSLES -ldl -llog -latomic -lm -latomic -lm && cd ."
C:/Dev/OpenAL/Alc/alu.cpp:0: error: undefined reference to 'void MixDirectHrtf_<NEONTag>(std::__ndk1::array<float, 1024ul>&, std::__ndk1::array<float, 1024ul>&, al::span<std::__ndk1::array<float, 1024ul> const, 18446744073709551615ul>, std::__ndk1::array<float, 2ul>*, DirectHrtfState*, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'float const* Resample_<LerpTag, NEONTag>(InterpState const*, float const*, int, int, float*, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'float const* Resample_<BSincTag, NEONTag>(InterpState const*, float const*, int, int, float*, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'void MixHrtfBlend_<NEONTag>(std::__ndk1::array<float, 1024ul>&, std::__ndk1::array<float, 1024ul>&, float const*, std::__ndk1::array<float, 2ul>*, int, int, HrtfParams const*, MixHrtfParams*, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'void MixHrtf_<NEONTag>(std::__ndk1::array<float, 1024ul>&, std::__ndk1::array<float, 1024ul>&, float const*, std::__ndk1::array<float, 2ul>*, int, int, MixHrtfParams*, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'void Mix_<NEONTag>(float const*, al::span<std::__ndk1::array<float, 1024ul>, 18446744073709551615ul>, float*, float const*, int, int, int)'
C:/Dev/OpenAL/Alc/mixvoice.cpp:0: error: undefined reference to 'void MixRow_<NEONTag>(std::__ndk1::array<float, 1024ul>&, float const*, al::span<std::__ndk1::array<float, 1024ul> const, 18446744073709551615ul>, int, int)'
```

**Comment by Slion** [comment]
Are the build machines not building for Android? Should I use a release branch rather than master?
I was secretly hoping compiling for Android would be easier than that :)
Is there some Android build instructions I could use?

**Comment by alexey-lysiuk** [comment]
[Here](https://github.com/kcat/openal-soft/blob/6ee49cad522b5bd12d954c508bae977e3c98cb07/.travis.yml#L54-L59) is how Android target is configured for Travis CI builds, and [here](https://github.com/kcat/openal-soft/blob/6ee49cad522b5bd12d954c508bae977e3c98cb07/.travis.yml#L27-L38) is the corresponding SDK setup.

**Comment by kcat** [comment]
> Looks like I'm one step further but now I'm getting the following errors. CLang and NDK 19.2.

Looks like it's building expecting NEON support, but isn't including the NEON source (the corresponding object file isn't in the link command). Not quite sure how that would happen since `HAVE_NEON` is set at the same point `mixer_neon.cpp` is added to the project source list. If you're still having problems, can you provide the cmake output from a fresh build directory?

**Comment by Slion** [comment]
Thanks. It looks like I am one step further after adding the following missing configuration options:
```
'-DALSOFT_REQUIRE_OPENSL=ON',
'-DALSOFT_EMBED_HRTF_DATA=YES'
```
However I just hit another error, see below. Any chance you guys could try building using NDK r19c and the clang toolchain?
```
C:\Users\micro\AppData\Local\Android\sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=aarch64-none-linux-android21 --gcc-toolchain=C:/Users/micro/AppData/Local/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64 --sysroot=C:/Users/micro/AppData/Local/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/windows-x86_64/sysroot  -DAL_ALEXT_PROTOTYPES -DAL_BUILD_LIBRARY -Dprj_ARCH_64 -Dprj_PLATFORM_ANDROID -DMONITOR_THREAD_CPU_TIME -DNOMINMAX -DOpenAL_EXPORTS -DSYMBIAN_ENABLE_64_BIT_FILE_SERVER_API -DSYMBIAN_ENABLE_SPLIT_HEADERS -DSYMBIAN_FIX_TDESC_CONSTRUCTORS -DUNICODE -D_UNICODE -D__CPU_HAS_VFP -D__DLL__ -D__EABI__ -D__GCC32__ -D__LEAVE_EQUALS_THROW__ -D__MARM__ -D__PLACEMENT_NEW_INLINE -D__PLACEMENT_VEC_NEW_INLINE -D__PLATSEC_UNLOCKED__ -D__SUPPORT_CPP_EXCEPTIONS__ -D__SYMBIAN_STDCPP_SUPPORT__ -D__SYMC__ -D__TOOLS__ -D__USE_VFP_MATH -D__WINS__ -IC:/Dev/VisualStudio/prj/Subs/OpenAL/include -IC:/Dev/bin/prj/Android/prj/Subs/OpenAL -IC:/Dev/VisualStudio/prj/Subs/OpenAL/Alc -IC:/Dev/VisualStudio/prj/Subs/OpenAL/OpenAL32/Include -IC:/Dev/VisualStudio/prj/Subs/OpenAL/common -std=c++11 -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -Wa,--noexecstack -Wformat -Werror=format-security -stdlib=libc++ -fms-extensions -O0 -fno-limit-debug-info  -fPIC   -Winline -Wall -Wextra -fno-math-errno -fvisibility=hidden -pthread -std=gnu++1z -MD -MT C:/Dev/bin/prj/Android/prj/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/fshifter.cpp.o -MF C:\Dev\bin\prj\Android\prj\Subs\OpenAL\CMakeFiles\OpenAL.dir\Alc\effects\fshifter.cpp.o.d -o C:/Dev/bin/prj/Android/prj/Subs/OpenAL/CMakeFiles/OpenAL.dir/Alc/effects/fshifter.cpp.o -c C:/Dev/VisualStudio/prj/Subs/OpenAL/Alc/effects/fshifter.cpp
In file included from C:/Dev/VisualStudio/prj/Subs/OpenAL/Alc/effects/fshifter.cpp:29:
In file included from C:/Dev/VisualStudio/prj/Subs/OpenAL/OpenAL32/Include\alMain.h:30:
In file included from C:/Dev/VisualStudio/prj/Subs/OpenAL/common\alnumeric.h:9:
In file included from C:\Users\micro\AppData\Local\Android\sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\lib64\clang\8.0.2\include\xmmintrin.h:27:
C:\Users\micro\AppData\Local\Android\sdk\ndk-bundle\toolchains\llvm\prebuilt\windows-x86_64\lib64\clang\8.0.2\include\mmintrin.h:47:5: error: use of undeclared identifier '__builtin_ia32_emms'; did you mean '__builtin_isless'?
    __builtin_ia32_emms();
```

**Comment by scmcom** [comment]
is there any particular reason you want to do this via AS? i have found it to be difficult if not impossible to do what you are attempting in AS, whereas with a cmake toolchain (on a unix system) compiling for android is basically trivial. i have described my method for doing so my most recent comment above, and have linked the toolchain in that comment. i am certain you will save yourself a lot of hassle with said approach :)

**Comment by Slion** [comment]
> is there any particular reason you want to do this via AS?

It's the official tool chain for Android. We have a large CMake C++17 project that's being built just fine through AS/gradle.

> i have found it to be difficult if not impossible to do what you are attempting in AS

A few years ago I would agree with that but Android Studio has come a long way since.

> whereas with a cmake toolchain (on a unix system) compiling for android is basically trivial.

As our targets are Windows and Android it makes a lot of sense to develop on Windows and cross compile from there. I seldom find an open source project that's compiling out-of-the-box even though the build machines are all green.

> i have described my method for doing so my most recent comment above

Thanks for the pointer. I'll make sure I take look at it.

**Comment by Slion** [comment]
I have a successful build if I disable the ARM targets.
It looks like an ARM configuration issue.
For some reason both `HAVE_EMMINTRIN_H` and `HAVE_SSE_INTRINSICS` are set even though they should not. I'm guessing that CMake magic we are doing here is not 100% cross compiler safe, or indeed AS is doing something funny. Off for a swim :)

**Comment by Slion** [comment]
Adding the following after the headers check fixed my build.
```
if(ANDROID_ABI MATCHES "arm")
    set(HAVE_XMMINTRIN_H 0) 
endif()
```

**Comment by Slion** [comment]
To sum it up, Android Studio build is working fine with the following `externalNativeBuild.cmake.arguments`, making sure you adjust the native tools path:

```
'-DALSOFT_NO_CONFIG_UTIL=ON',
'-DALSOFT_EXAMPLES=OFF',
'-DALSOFT_UTILS=OFF',
'-DALSOFT_TESTS=OFF',
'-DALSOFT_NATIVE_TOOLS_PATH=C:/Path/To/bin/Prj/Subs/OpenAL/native-tools',
'-DALSOFT_REQUIRE_OPENSL=ON',
'-DALSOFT_EMBED_HRTF_DATA=YES'
```
Providing that you applied that patch:
```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -389,6 +389,11 @@ CHECK_INCLUDE_FILE(pmmintrin.h HAVE_PMMINTRIN_H "${SSE3_SWITCH}")
 CHECK_INCLUDE_FILE(smmintrin.h HAVE_SMMINTRIN_H "${SSE4_1_SWITCH}")
 CHECK_INCLUDE_FILE(arm_neon.h HAVE_ARM_NEON_H "${FPU_NEON_SWITCH}")
 
+if(ANDROID_ABI MATCHES "arm")
+    set(HAVE_XMMINTRIN_H 0) 
+endif()
+
+
 SET(SSE_FLAGS )
 SET(FPMATH_SET "0")
 IF(CMAKE_SIZEOF_VOID_P MATCHES "4")
@@ -1180,8 +1185,9 @@ ELSE()
         VERBATIM
     )
 ENDIF()
+
 ADD_CUSTOM_TARGET(native-tools
-    DEPENDS "${BIN2H_COMMAND}" "${BSINCGEN_COMMAND}"
+#    DEPENDS "${BIN2H_COMMAND}" "${BSINCGEN_COMMAND}"
     VERBATIM
 )
```

It would be great if we could find a proper solution to contribute to master.

**Comment by Slion** [comment]
Here is what the cmake output looks like when building for ARM with my patch:
```
Could NOT find ALSA (missing: ALSA_LIBRARY ALSA_INCLUDE_DIR) 
Could NOT find OSS (missing: OSS_INCLUDE_DIR) 
Could NOT find AudioIO (missing: AUDIOIO_INCLUDE_DIR) 
Could NOT find SoundIO (missing: SOUNDIO_LIBRARY SOUNDIO_INCLUDE_DIR) 
Could NOT find QSA (missing: QSA_LIBRARY QSA_INCLUDE_DIR) 
Could NOT find PortAudio (missing: PORTAUDIO_LIBRARY PORTAUDIO_INCLUDE_DIR) 
Could NOT find PulseAudio (missing: PULSEAUDIO_LIBRARY PULSEAUDIO_INCLUDE_DIR) 
Could NOT find JACK (missing: JACK_LIBRARY JACK_INCLUDE_DIR) 
Could NOT find SDL2 (missing: SDL2_LIBRARY SDL2_INCLUDE_DIR) 
Building OpenAL with support for the following backends:
     OpenSL, WaveFile, Null
Building with support for CPU extensions:
    Default, Neon
Embedding HRTF datasets
Installing sample configuration
Installing HRTF definitions
Installing AmbDec presets
```

**Comment by kcat** [comment]
Can you provide the full CMake output from a fresh build without your changes? The Travis CI build shows it properly failing to find/use the SSE headers, so it's working there. It may also be useful to look at the contents of `CMakeFiles/CMakeOutput.txt`, to see why it's successfully finding and using SSE headers and intrinsics there but then failing with it during the actual build.

The issue with `BIN2H_COMMAND` and `BSINCGEN_COMMAND` I suspect is due to the `.exe` extension Windows has for native executables. I've pushed a potential fix for that.

**Comment by Slion** [comment]
I did pull your latest changes and made a clean build.
The issue with native tools dependency is solved.

Here are the logs you asked for:
[cmake_build_outpu-arm64-v8a.txt](https://github.com/kcat/openal-soft/files/3244570/cmake_build_outpu-arm64-v8a.txt)
[cmake_build_output-armeabi-v7a.txt](https://github.com/kcat/openal-soft/files/3244572/cmake_build_output-armeabi-v7a.txt)
[cmake_build_output-x86.txt](https://github.com/kcat/openal-soft/files/3244574/cmake_build_output-x86.txt)
[cmake_build_output-x86_64.txt](https://github.com/kcat/openal-soft/files/3244579/cmake_build_output-x86_64.txt)

**Comment by Slion** [comment]
Looking at the CMake code and output it seems that `HAVE_SSE_INTRINSICS` should not be defined. Yet it must be since `xmmintrin.h` is included from `alnumeric.h` which causes the compile error.

Wondering if that could be a good old uninitialised variable issue I went ahead and added the following before the `CHECK_INCLUDE_FILE` block:

```
set(HAVE_XMMINTRIN_H 0)
set(HAVE_EMMINTRIN_H 0)
set(HAVE_PMMINTRIN_H 0)
set(HAVE_SMMINTRIN_H 0) 
set(HAVE_ARM_NEON_H 0) 
```

Sure enough that fixed it.
I'm using CMake 3.10.2 as provided by the NDK.
I'm not sure if we should report that to the Android NDK team or CMake guys or both.

**Comment by kcat** [comment]
> Here are the logs you asked for:

These all look correct. The ARM builds fail the `*mmintrin.h` header checks, which should prevent `HAVE_SSE_INTRINSICS` from getting defined, while the `x86`/`x86-64`builds find them and properly enable it. Can you post the full compile output of one of those ARM builds that have the problem?

> Wondering if that could be a good old uninitialised variable issue I went ahead and added the following before the `CHECK_INCLUDE_FILE` block:
> ...
> Sure enough that fixed it.

That effectively disables the checks and forces it to assume those headers aren't available. `CHECK_INCLUDE_FILE` doesn't do anything if the result variable is already defined (if it already checked, the result is cached and it won't check again regardless if it was found or not). Are you sure you started in a clean build directory for the ARM builds? A stale cache from an x86 build could certainly make it think those headers are still available for an ARM build, and erroneously enable SSE intrinsics when they're not actually usable.

**Comment by Slion** [comment]
> A stale cache from an x86

Not sure how that would happen. Each build variant has its own CMakeCache.txt in its own directory.
Looking at the content of the cache it looks consistent.

However it seems to be the problem here indeed. If I build only the ARM variants it works. If I try to build both x86 and ARM variants together it fails.

So it seems it is indeed a an issue with Android Studio/Ninja. To get to the bottom of that I should really try to reproduce the issue with a minimal project making use of OpenAL. Not sure I'll ever find the time to do that.

> Can you post the full compile output of one of those ARM builds that have the problem?

I rather not, not until I can isolate the problem as explained above. However there is no much to it outside the error logs posted above.
